docs(adr): ADR-005: Support multisig chain link (#634)
desmos-labs/desmos: docs/architecture/adr-005-support-multisig-chain-link.md

3. Desmos verifies the `Proof` object to guarantee that the user really owns such account and thus it can be linked successfully to their Desmos profile.

Currently, this process works properly for single-signature accounts, but it does not support multi-signature accounts. This is due to the fact that the `Proof` type only supports single-account's signatures, and its `Verify` function is only able to verify such signature type.

## Decision

We propose to change the `Proof#Signature` field to be of type [`SignatureDescriptor_Data`](https://github.com/cosmos/cosmos-sdk/blob/master/proto/cosmos/tx/signing/v1beta1/signing.proto#L57).

### Proof implementation

`SignatureDescriptor_Data` supports storing both single and multi-sig signatures. Moreover, the Cosmos SDK provides a [function](https://github.com/cosmos/cosmos-sdk/blob/master/types/tx/signing/signature.go#L65) to convert such type into the corresponding interface ([`SignatureData`](https://github.com/cosmos/cosmos-sdk/blob/master/types/tx/signing/signature_data.go#L10)), which is required from the `Verify` function in order to validate such signature. 

The verification process can be implemented as follows:

> 1. If it's a `SingleSignatureData`, make sure the account public key is a cryptotypes. `PubKey` and 
> then use the `VerifySignature` method to verify the signature.

Reviewer: ```suggestion
1. If it's a `SingleSignatureData`, make sure the account public key is a `cryptotypes.PubKey` and then use the `VerifySignature` method to verify the signature.
```

> 2. If it's a `MultiSignatureData`, make sure the account public key is a `multisig.PubKey` and 
> then use the `VerifyMultisignature` method to verify the signature.

Reviewer: ```suggestion
2. If it's a `MultiSignatureData`, make sure the account public key is a `multisig.PubKey` and then use the `VerifyMultisignature` method to verify the signature.
```

### CLI implementation

> We propose to use an interactive prompt to create a new chain-link JSON for both single signature and multi signature account. For the single signature account, the process will remain the same as before. For the multi signature account, we will add the new feature beyond the remained interactive prompt.

Reviewer: ```suggestion
We propose to use an interactive prompt to create a new chain-link JSON for both single signature and multi signature account. For the single signature account, the process will remain the same as before. For the multi signature account, we will add the new feature beyond the current ones.
```

> To simplify things for multisig accounts, we propose to create a chain link JSON starting from a transaction signed with a multisig account. Inside the Cosmos SDK, multisig account transactions signing depends on the `tx sign` and 
> `tx multisign` commands. To send a transaction to the node, the following process takes place: 

Reviewer: ```suggestion
To simplify things for multisig accounts, we propose to create a chain link JSON starting from a transaction signed with a multisig account. Inside the Cosmos SDK, multisig account transactions signing depends on the `tx sign` and `tx multisign` commands. To send a transaction to the node, the following process takes place: 
```

> 1. the multisig account owners create a raw transaction file

Reviewer: ```suggestion
1. the multisig account owners create a raw transaction file;
```

> 2. the threshold number of signers sign it to using the `tx sign` command

Reviewer: ```suggestion
2. the threshold number of signers sign it to using the `tx sign` command;
```

	}

	txFactory := tx.NewFactoryCLI(clientCtx, cmd.Flags())
	if txFactory.SignMode() == signingtypes.SignMode_SIGN_MODE_UNSPECIFIED {
		txFactory = txFactory.WithSignMode(signingtypes.SignMode_SIGN_MODE_LEGACY_AMINO_JSON)
	}

	sigs, err := txBuilder.GetTx().GetSignaturesV2()
	if len(sigs) != 1 {
		return profilescliutils.ChainLinkJSON{}, fmt.Errorf("invalid number of signatures")
	}
	multisigSig := sigs[0]

	signingData := authSigning.SignerData{
		ChainID:       chainID,
		AccountNumber: txFactory.AccountNumber(),
		Sequence:      txFactory.Sequence(),
	}
	// the bytes of plain text
	value, err := txCfg.SignModeHandler().GetSignBytes(txFactory.SignMode(), signingData, txBuilder.GetTx())
	if err != nil {
		return profilescliutils.ChainLinkJSON{}, err
	}

	addr, _ := sdk.Bech32ifyAddressBytes(chain.Prefix, multisigSig.PubKey.Address().Bytes())
	return profilescliutils.NewChainLinkJSON(
		profilestypes.NewBech32Address(addr, chain.Prefix),
		profilestypes.NewProof(multisigSig.PubKey, signingtypes.SignatureDataToProto(multisigSig.Data), hex.EncodeToString(value)),
		profilestypes.NewChainConfig(chain.Name),
	), nil
}
```

## Consequences

### Backwards Compatibility

> Since we change the `Proof#Signature` field from a hex-encode string into a `SignatureDescriptor_Data`, this feature is not backwards compatible. Migrating the old chain link signatures to `SignatureDescriptor_Data`
> is required.

Reviewer: ```suggestion
Since we change the `Proof#Signature` field from a hex-encoded string into a `SignatureDescriptor_Data`, this feature is not backwards compatible. For this reason, it will be needed to migrate the old chain link signatures to `SignatureDescriptor_Data`.
```
